feat: add findNextNOccurrences and findNextTilDateOccurrences

VittorioParagallo opened the pull request:
This PR adds a `findNextNOccurrences` and `findNextTilDateOccurrences` to the `schedule` class.

Commit:
added findNextNOccurrences and findNextTillDateOccurrences to schedule class

andyhorn commented on lib/src/schedules.dart:
```diff
@@ -35,6 +35,55 @@ sealed class Schedule {
     return true;
   }
+
+  List<DateTime> findNextNOccurrences(Schedule schedule, int n,
```
Do we need to pass in a `Schedule`? Can we just use the current instance?

This same question applies to `findNextTilDateOccurrences`, too.

VittorioParagallo replied:
Yes is possible to use the current instance, i just added as a static methos to be as less invasive as possible. I'm a bit busy these days, but if you like more that solution, i'll write the code as soon as i find some time. Then in the future would be nice to calculate next occurence instead of "brute forcing" with every single day try.

andyhorn replied:
I hear ya. No rush!

I agree that calculating instead of brute-forcing it would be a nice improvement. Once this is merged, I might play around with the idea.

andyhorn commented on lib/src/schedules.dart:
```diff
@@ -35,6 +35,55 @@ sealed class Schedule {
+    final dates = <DateTime>[];
+    var currentDate = fromDate ?? DateTime.now();
+    excludeDates = excludeDates
+        .map((element) => DateTime(element.year, element.month, element.day))
```
I have an extension to strip the time off of a `DateTime` object in the `extensions.dart` file. We could probably make it a public extension - what do you think?

andyhorn replied:
Or perhaps make it a separate utility function that is only available within the package?

VittorioParagallo replied:
> I have an extension to strip the time off of a `DateTime` object in the `extensions.dart` file. We could probably make it a public extension - what do you think?

I saw it, but yes, it is private. Not sure about making it a public extension we should think a bit about that.

andyhorn replied:
I think we can make it a public extension. The extensions are not exported from the package, so I don't see any reason not to 👍🏻

Commit:
removed the time data in current dates

Commit:
prepared unit test for a bug in findNoccurences
because when the n'th occurrence goes after the
end date the method goes in infinite loop

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
fixed finNoccurences bug and implemented unit testing

VittorioParagallo commented:
i think it could be merged

Commit:
removed the schedule parameter from findNextNOccurrences and findNextTillDateOccurrences as it was left since it was a static method

VittorioParagallo commented:
I don't understand what should i do next to merge it

andyhorn committed:
Merge branch 'main' into main

Commit:
corrected errors not allowing to pull

andyhorn commented:
Sorry, been out of the game for awhile with a new kid; just getting back to this!

andyhorn commented on test/schedules_test.dart:
```diff
@@ -263,4 +263,21 @@ void main() {
+          frequency: 1,
+          weekdays: [2]);
+
+      List<DateTime> listDates = schedule.findNextNOccurrences( 33);
```
This test is now failing because the `findNextNOccurrences` method is defaulting to `today`, which is now beyond the `endDate`, resulting in 0 items.

If you set the `fromDate` equal to the `schedule.startDate`, it should pass.

Commit:
corrected tests

VittorioParagallo commented:
done thank you

Il giorno gio 17 ott 2024 alle ore 20:07 Andy ***@***.***> ha
scritto:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In test/schedules_test.dart
> <https://github.com/andyhorn/schedules/pull/7#discussion_r1805214627>:
>
> > @@ -263,4 +263,21 @@ void main() {
>        });
>      });
>    });
> +
> +  group('findNextNOccurrences', () {
> +    test('''given n required occurences if the end date happens
> +      before the n'th occurence the method returns a lower number
> +      of occurences within the start-date and end-date constraint''', () {
> +      Schedule schedule = Weekly(
> +          endDate: DateTime.parse('2024-09-21 00:00:00.000Z'),
> +          startDate: DateTime.parse('2024-05-20 00:00:00.000Z'),
> +          frequency: 1,
> +          weekdays: [2]);
> +
> +      List<DateTime> listDates = schedule.findNextNOccurrences( 33);
>
> This test is now failing because the findNextNOccurrences method is
> defaulting to today, which is now beyond the endDate, resulting in 0
> items.
>
> If you set the fromDate equal to the schedule.startDate, it should pass.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/andyhorn/schedules/pull/7#pullrequestreview-2375940745>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AN3UJSD7SOTURK7JTW7PHTTZ374FLAVCNFSM6AAAAABHS7URC6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGNZVHE2DANZUGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

VittorioParagallo commented:
Up
